src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs

```diff
@@ -557,77 +1397,115 @@ public long LocalCount
             {
-                //
-                // Set up our thread-local data
-                //
-                // Use operate on workQueue local to try block so it can be enregistered
-                ThreadPoolWorkQueue workQueue = outerWorkQueue;
-                ThreadPoolWorkQueueThreadLocals tl = workQueue.GetOrCreateThreadLocals();
-                Thread currentThread = tl.currentThread;
-
+                Thread currentThread = Thread.CurrentThread;
                 // Start on clean ExecutionContext and SynchronizationContext
                 currentThread._executionContext = null;
                 currentThread._synchronizationContext = null;
 
+                // Every new worker that finds work will ask for parallelizm increase, but only once.
+                // This helps with front-edge ramping up from cold states.
+                bool requestedAnotherWorker = false;
+
+                //
+                // Use operate on workQueue local to try block so it can be enregistered
+                ThreadPoolWorkQueue workQueue = ThreadPoolGlobals.workQueue;
+
                 //
                 // Loop until our quantum expires or there is no work.
                 //
-                while (ThreadPool.KeepDispatching(startTickCount))
+                do
                 {
-                    bool missedSteal = false;
-                    // Use operate on workItem local to try block so it can be enregistered
-                    object? workItem = workQueue.Dequeue(tl, ref missedSteal);
+                    var localQueue = workQueue.GetOrAddLocalQueue();
+                    object? workItem = localQueue.TryPop();
 
                     if (workItem == null)
                     {
-                        //
-                        // No work.
-                        // If we missed a steal, though, there may be more work in the queue.
-                        // Instead of looping around and trying again, we'll just request another thread.  Hopefully the thread
-                        // that owns the contended work-stealing queue will pick up its own workitems in the meantime,
-                        // which will be more efficient than this thread doing it anyway.
-                        //
-                        needAnotherThread = missedSteal;
+                        var spinner = new SpinWait();
+                        for (; ; )
+                        {
+                            // we could not pop, try stealing
+                            bool missedSteal = false;
+                            workItem = workQueue.DequeueAny(ref missedSteal, localQueue);
+                            if (workItem != null)
+                            {
+                                break;
+                            }
+
+                            // if there is no more work, start reducing parallelism exponentially
+                            if (!missedSteal && (localQueue.NextRnd() & 1) == 0)
```

Comment:
Why not just return when finding that there are no items to pop or steal? I don't see an obvious motivation here, there should at least be a more informative comment on why this is being done. I figure there may be some perf advantage to spin-waiting over returning, but this would not be for free, so wondering about the reason.

Reply:
I see that `TryRemove` does not retry on contention (same as before) and so this is a retry over all queues without having to leave and come back to `Dispatch`. Main difference is on the retry this version skips the local queue this thread is currently assigned to, and any unmanaged work items that may be queued on the native side. The global queue and perhaps many local queues would have been empty before a `TryRemove` failure, so I'm not sure whether it would be better to retry on failed contention in `TryRemove` or to retry all queues. If stealing because it's the end of a burst of work items, then it would probably make sense to retry on contention. If stealing because another local queue has more incoming work items or progress is slow on that queue, then it still feels like it would be better to retry on contention. If while stealing (after seeing the global queue is empty) there are new global work items being added then it would make sense to retry both, but not sure that would be common.